brainiak/matnormal/covs.py

```diff
@@ -343,49 +357,43 @@ def __init__(self, size=None, Sigma=None):
 
         super(CovUnconstrainedCholesky, self).__init__(size)
 
+        # number of parameters in the triangular mat
+        npar = (size * (size + 1)) // 2
+
         if Sigma is None:
-            self.L_full = tf.Variable(
-                tf.random.normal([size, size], dtype=tf.float64),
-                name="L_full",
-                dtype="float64",
+            self.L_flat = tf.Variable(
+                tf.random.normal([npar], dtype=tf.float64), name="L_flat"
             )
 
         else:
-            # in order to respect the Sigma we got passed in, we log the diag
-            # which we will later exp. a little ugly but this
-            # is a rare use case
             L = np.linalg.cholesky(Sigma)
-            L[np.diag_indices_from(L)] = np.log(np.diag(L))
-            self.L_full = tf.Variable(L, name="L_full", dtype="float64")
-
-        # Zero out triu of L_full to get cholesky L.
-        # This seems dumb but TF is smart enough to set the gradient to zero
-        # for those elements, and the alternative (fill_lower_triangular from
-        # contrib.distributions) is inefficient and recommends not doing the
-        # packing (for now).
-        # Also: to make the parameterization unique we exp the diagonal so
-        # it's positive.
-
-        L_indeterminate = tf.linalg.band_part(self.L_full, -1, 0)
-        self.L = tf.linalg.set_diag(
-            L_indeterminate, tf.exp(tf.linalg.diag_part(L_indeterminate))
-        )
+            self.L_flat = tf.Variable(
+                flatten_cholesky_unique(L), name="L_flat")
+
+        self.optimize_vars = [self.L_flat]
 
     @property
-    def logdet(self):
+    def L(self):
+        """
+        Cholesky factor of this covariance
+        """
+        return unflatten_cholesky_unique(self.L_flat)
 
-        # We save a log here by using the diag of L_full
-        return 2 * tf.reduce_sum(input_tensor=(tf.linalg.diag_part(self.L_full)))
+    @property
+    def logdet(self):
+        return 2 * tf.reduce_sum(input_tensor=tf.math.log(
+                                 tf.linalg.diag_part(self.L)))
```

Comment:
One question: is there any possibility that the diagonal values are negative? The covariance can still be positive definite if L has negative diagonal value. How about you take the square of the diagonal part before log and remove the multiplication by 2 outside?

Reply:
`self.L` is a property that is computed by running `unflatten_cholesky_unique(self.L_flat)` (line 376), defined on line 74 of `utils.py`, as such:
```
    L = tfp.math.fill_triangular(L_flat)
    # exp diag for unique parameterization
    L = tf.linalg.set_diag(L, tf.exp(tf.linalg.diag_part(L)))
```
So basically, I exp the diagonal to make it positive, which makes the parameterization unique and valid (I believe). 

I'm open to another operation that makes it positive (like squaring) if you think that's a better idea, but I prefer not to remove the factor of 2 outside even is squaring the diag would be mathematically valid -- I think operating in log space will generally be more numerically stable, and having it in the `logdet` method makes it easy to understand what's going on here (since it looks like a textbook formula).

Reply:
Agree. Sorry I forget you had this way to ensure unique parametrization. So this is fine.